instance/management/commands/migrate_swift_to_s3.py

```diff
@@ -0,0 +1,135 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Migrate SWIFT assets to S3.
+One-time script.
+"""
+
+# Imports #####################################################################
+
+import re
+import subprocess
+import logging
+
+from django.core.management.base import BaseCommand
+
+from instance.models.openedx_instance import OpenEdXInstance
+import swiftclient
+
+LOG = logging.getLogger(__name__)
+
+# Classes #####################################################################
+
+
+class Command(BaseCommand):
+    """
+    Migrate SWIFT to S3.
+    """
+    help = (
+        'Migrate all servers\' data from SWIFT to S3.'
+    )
+
+    def __init__(self, *args, **kwargs):
+        super(Command, self).__init__(*args, **kwargs)
+        self.options = {}
+        self.retried = {}
+
+    def _get_swift_connection(self, instance):
+        return swiftclient.Connection(
+            user=instance.swift_openstack_user,
+            key=instance.swift_openstack_password,
+            authurl=instance.swift_openstack_auth_url,
+            tenant_name=instance.swift_openstack_tenant,
+            auth_version='2',
+            os_options={'region_name': instance.swift_openstack_region}
+        )
+
+    def _create_rclone_config(self, access_key, secret_key):
+        """
+        Creates rclone config, to ease the process of copying
+        """
+        rclone_config = "rclone config create ocim-s3 s3 access_key_id %s secret_access_key %s" % (access_key, secret_key)
+        print("Would run this command: %s.       ← FIXME test it and run it" % rclone_config)
+        return
+
+        subprocess.Popen(rclone_config % (access_key, secret_key), shell=True)
+
+
+    def _copy_subdirectory(self, key_name, new_name, swift_container, s3_bucket):
+        """
+        Copy subdirectory using rclone
+        """
+        LOG.info("%s ---> %s" % (key_name, new_name))
+        command = "rclone copy ocim-swift:%s/%s ocim-s3:%s/%s" % (swift_container, key_name, s3_bucket, new_name)
+        print("Would run this command: %s.       ← FIXME test it and run it" % command)
+        return
+        subprocess.Popen(command, shell=True)
+
+    def _migrate_swift_to_s3(self, instance):
+        # FIXME Will these 2 values already be stored in the instance? If not, where to get them from? (Or should we set them and save?)
```

Comment:
you mean the iam keys and the s3 bucket? No, they will need to be created. You can do so by running:

```python
instance.s3_bucket_name = instane.bucket_name
instance.create_iam_user()
instance._create_bucket()
instance.save()
```

There has to be some wait time after creating the iam keys, before creating the bucket, cause the IAM keys don't get propagated immediately, you can add some seconds of sleep time here.

Also instance.storage_type has to be updated after doing the migration to s3, and a new appserver needs to be spawned for it to take the new bucket.

Reply:
@rocioar @smarnach is this script also responsible for spawning a new server?
In that case, the script should be run carefully, and do much better error handling, and use some tags etc.
And run either with few instances, or interactively („press enter to continue“). Or use tags that later can be used by instance_redeploy (in case nothing else needs to be done with the servers).

Yes, it makes sense to change .storage_type to s3; it seems the script should do this.